File: solutions/0206-reverse-linked-list/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        if(head==None):
            return head
        if(head.next==None):
            return head
        if(head.next.next ==None):
            temp = head.next 
            temp.next = head
            head.next = None
            return temp
        
        P = head.next 
        Q = head.next.next
        head.next = None
        while(Q!=None):
            P.next = head
            head = P
            P = Q
            Q = Q.next
        P.next = head
        return P
        
            
    
            
            
            
            
            
            
        
        

File: solutions/0206-reverse-linked-list/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if(!head or !head->next) return head;
        ListNode* prev  = head;
        ListNode* curr = head->next;
        ListNode* nex = head->next->next;
        head->next = nullptr;
        while(nex!=nullptr){
            curr->next = prev;
            prev = curr;
            curr = nex;
            nex = nex->next;
        }
        curr->next = prev;
        return curr;
    }
};
